util/conversion/conversion.go

```diff
@@ -30,13 +35,50 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/apimachinery/pkg/util/json"
 	"k8s.io/utils/diff"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/conversion"
 )
 
 const (
 	DataAnnotation = "cluster.x-k8s.io/conversion-data"
 )
 
+var (
+	contract = clusterv1.GroupVersion.String()
+)
+
+// ConvertReferenceAPIContract takes a client and object reference, queries the API Server for
+// the Custom Resource Definition and looks which one is the stored version available.
+//
+// The object passed as input is modified in place if an updated compatible version is found.
+func ConvertReferenceAPIContract(ctx context.Context, c client.Client, ref *corev1.ObjectReference) error {
+	// TODO(vincepri): Use CRDv1 once we have support for it.
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	key := client.ObjectKey{Name: ref.Kind}
+	if err := c.Get(ctx, key, crd); err != nil {
+		return errors.Wrapf(err, "cannot find CustomResourceDefinition for %v", ref.GroupVersionKind())
+	}
+
+	// If there is no label, return early without changing the reference.
+	supportedVersions, ok := crd.Labels[contract]
+	if !ok {
+		return nil
+	}
+
+	// If the label is present, but there list is empty, return an error.
+	versions := strings.Split(supportedVersions, ",")
+	if len(versions) == 0 {
+		return errors.Errorf("cannot find any version matching contract %q for CRD %v", contract, crd.Name)
+	}
+
+	// Modify the GroupVersionKind with the new version.
+	gvk := ref.GroupVersionKind()
+	gvk.Version = versions[len(versions)-1]
```

Comment:
I'm wondering if it would make sense to sort the available versions using https://godoc.org/k8s.io/apimachinery/pkg/version#CompareKubeAwareVersionStrings?

Reply:
Ah nice! I can definitely add this in